File: backend/schemas/reel.py
"""Pydantic schemas for Reel-related requests and responses."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, HttpUrl


class ReelCreate(BaseModel):
    url: HttpUrl = Field(..., description="The Instagram reel URL")


class ReelResponse(BaseModel):
    id: UUID
    url: str
    status: str
    category_id: Optional[UUID] = None
    creator_handle: Optional[str] = None
    thumbnail_url: Optional[str] = None
    transcript: Optional[str] = None
    caption: Optional[str] = None
    hashtags: list[str] = []
    summary: Optional[str] = None
    confidence: Optional[float] = None
    has_audio: Optional[bool] = None
    retry_count: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class CategoryChoiceRequest(BaseModel):
    """Body for PATCH /reels/{reel_id}/category.

    category_name: exact category name (string) → assigns category, marks ready.
    category_name: null → skips to uncategorised immediately.
    """
    category_name: Optional[str] = None

File: frontend/TempReels/Reel.swift
import Foundation

struct CategoryInfo: Decodable, Hashable {
    let name: String
}

struct Reel: Identifiable, Decodable, Hashable {
    let id: UUID
    let userId: UUID
    let categoryId: UUID?
    let url: String
    let creatorHandle: String?
    let thumbnailUrl: String?
    let transcript: String?
    let caption: String?
    let hashtags: [String]
    let summary: String?
    let confidence: Float?
    let hasAudio: Bool?
    let status: String
    let retryCount: Int?
    let deletedAt: Date?
    let createdAt: Date
    let updatedAt: Date
    let categories: CategoryInfo?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case categoryId = "category_id"
        case url
        case creatorHandle = "creator_handle"
        case thumbnailUrl = "thumbnail_url"
        case transcript
        case caption
        case hashtags
        case summary
        case confidence
        case hasAudio = "has_audio"
        case status
        case retryCount = "retry_count"
        case deletedAt = "deleted_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case categories
    }
}
